### ISSUE TITLE
Elixir checker executes code in buffer

### ISSUE THREAD

**Issue by mattly** [flycheck/flycheck#630]
While investigating the issue that led to #628, [Elixir creator José Valim mentioned that elixirc executes code](https://github.com/elixir-lang/elixir/issues/3282#issuecomment-94381550). I discovered this while editing a cpu-intensive `.exs` script and noticing that the script was actually being ran, but José says that the `.exs` extension differs from the main `.ex` extension in that the only real difference is one of convention – elixir code is evaluated by `elixirc` on compile.

So that means you can create a new elixir buffer containing:
``` elixir
File.write("foo", "bar")
```
and it will write a file to disk. You could also do `File.rm_rf!` or any number of things. While the majority of elixir code I work on is my own and I can trust it, I cannot trust any code I open in Emacs is going to be safe, nor should I have to.

The solution is not to use `elixirc` for syntax checking, but converting flycheck over will not prove simple, and will not provide actual compilation errors that can only be gained by evaluating and compiling the code.

Since I'm still new to Emacs and its lisp variant, I'm going to sketch out the issue and its solution first, and then set about solving it.

This is a safe way to syntax-check elixir code:
```
elixir -e 'r = System.argv |> List.first |> File.read! |> Code.string_to_quoted; if elem(r, 0) == :error do IO.inspect(elem(r,1)); end' -- filename.ex
```
It's perhaps not the most idiomatic elixir code, but I was aiming for something that would work on the command-line instead of from a file. It also outputs errors as elixir s-expressions, for example `{6, "missing terminator: ) (for \"(\" starting at line 1)", ""}` instead of
```
== Compilation error on file invalid_ex.ex ==
** (TokenMissingError) invalid_ex.ex:6: missing terminator: ) (for "(" starting at line 1)
    (elixir) lib/kernel/parallel_compiler.ex:95: anonymous fn/4 in Kernel.ParallelCompiler.spawn_compilers/8
```

Finally, this method only checks for syntax errors, it will not generate compile-time warnings or errors such as missing modules, invalid arguments, typos in keywords, etc. The only way to do that is to evaluate the code. I would *love* to have a mode in flycheck that continued to provide that level of checking, but only for trusted projects. I'm not sure there's a way to flip between "safe, shallow checker" and "unsafe, deep checker" however.

**Comment by lunaryorn** [comment]
@mattly  I don't think that I want to be bothered with this complexity.  I'll rather remove the Elixir syntax checker entirely than introducing additional complexity just to work around a compiler which behaves badly.  It's not my fault that it's broken, and it's not my issue to fix.

For now, if you frequently work with untrusted Elixir code, remove the checker from `flycheck-checkers`, and manually enable it for trusted code with `M-x flycheck-select-checker`.

If there's no other way, I'll remove the checker this weekend.

**Comment by mattly** [comment]
I believe the compiler does the evaluation because it has to handle macros at compile-time.

This makes me sad, but I totally understand. On the flip side, maybe I'll create a simple CLI tool available via homebrew or such that will do safe syntax-checking and help fill a niche.

**Comment by lunaryorn** [comment]
@mattly I've now removed this syntax checker.  Feel free to maintain your own Flycheck extension with a limited but safe checker.  I'll help you as far as I can.
